slices/libpam-runtime.yaml

```diff
@@ -18,40 +18,133 @@ slices:
+            d = []
+            for j in range(i + 1, len(lines)):
+              if ":" in lines[j]:
+                break
+              d.append(lines[j])
+            confdata[m] = {t: {p: d}}
+
+      def reconfigure_data_line(section, line, i):
+        res = ""
+        upd = line.replace("success=end", "success=" + str(i))
+        if section == "session-noninteractive":
+          res += "session" + upd + "\n"
+        else:
+          res += section + upd + "\n"
+        return res
+
+      def build_block(data, section, block, existing):
+        si = 1
+        res = existing
+        if section in data:
+          types = data[section]
+          if block in types:
+            # get the keys (which are the priorities) and
+            # then reverse sort them to get the highest priority
+            # first
+            itemsByPriority = types[block]
+            keys = itemsByPriority.keys()
+            keys = sorted(keys, reverse=True)
+            for p in keys:
+              for d in itemsByPriority[p]:
+                res = reconfigure_data_line(section, d, si) + "\n"
+                si += 1
+
+        # no primary block, so output a stock pam_permit line
+        # to keep the stack intact
+        if res == "" and block == "Primary":
+          return reconfigure_data_line(section, "\t[default=1]\t\t\tpam_permit.so\n", 1);
+        return res
 
-  # Used by pam-auth-update to generate the pam.d/ config
-  # files inside etc. However pam-auth-update relies on
-  # debconf to actually work and discover these packages.
+      fnames = ["account", "auth", "password", "session", "session-noninteractive"]
+      idnames = ["$account", "$auth", "$password", "$session", "$session_nonint"]
+
+      for i in range(len(fnames)):
+        fn = fnames[i]
+        template = content.read("/usr/share/pam/common-" + fn)
+        pb = build_block(confdata, fn, "Primary", "")
+        ab = build_block(confdata, fn, "Additional", "")
+
+        # session also includes settings from the session-noninteractive,
+        # but not the other way around
+        if fn == "session":
+          pb = build_block(confdata, "session-noninteractive", "Primary", pb)
+          ab = build_block(confdata, "session-noninteractive", "Additional", ab)
+
+        template = template.replace(idnames[i] + "_primary", pb)
+        template = template.replace(idnames[i] + "_additional", ab)
+        content.write("/etc/pam.d/common-" + fn, template)
+
+  # the following two slices are kept for backwards-compatibility
+  # however the pam-defaults have been modified to deliver correct
+  # files instead of templates
   pam-config:
     contents:
       /usr/share/pam-configs/unix:
 
-  # default templates for the /etc/pam.d files that are used by
-  # pam-auth-update to generate the /etc/pam.d versions based on
-  # additional plugs in /usr/share/pam-configs/*.
   pam-defaults:
+    essential:
+      - libpam-runtime_config
```

Comment:
Why though? It feels like `pam-defaults` should deliver the defaults, which are the templates. Since you're not depending on `pam-defaults` anywhere, I suggest keeping it as is to avoid regressions.

Reply:
For clarification: files in `/usr/share/pam/` are just template files that are used by pam-auth-update to generate `/etc/pam.d/common-*` files. Since the mutate script is mimicking that, there is no need to keep them anymore.

I don't see any of our slices referencing `pam-defaults`, so it may be fine to remove those files from it, but there is some small risk that someone is using that slice in a container image to run `pam-auth-update` with it.

Reply:
@linostar i don't think that's an issue, there was no slice delivering pam-uath-update prior to this?

Reply:
Since we are keeping the `pam-defaults` slice in the fear that someone could still be using it, it seems reasonable to me to not change the contents it had. `config` provides the new contents anyway.

Reply:
let's keep the pam-defaults as is to avoid regressions. And config can indeed deliver the mutated defaults and resulting configs